src/lib/context.js

```diff
@@ -1,134 +1,97 @@
+      .forEach(res => this.bundle.addMessages(res));
   }
 
-  _resolve(langs, keys, formatter, prevResolved) {
-    const lang = langs[0];
-
-    if (!lang) {
-      return reportMissing.call(this, keys, formatter, prevResolved);
-    }
-
-    let hasUnresolved = false;
-
-    const resolved = keys.map((key, i) => {
-      if (prevResolved && prevResolved[i] !== undefined) {
-        return prevResolved[i];
-      }
+  _formatKeys(keys, method) {
+    return keys.map((key, i) => {
       const [id, args] = Array.isArray(key) ?
         key : [key, undefined];
-      const entity = this._getEntity(lang, id);
 
-      if (entity) {
-        return formatter.call(this, lang, args, entity);
-      }
-
-      hasUnresolved = true;
+      // XXX Context should handle errors somehow; emit/return?
+      const [result] = method.call(this, id, args);
+      return result;
     });
+  }
 
-    if (!hasUnresolved) {
-      return resolved;
+  formatValue(id, args) {
+    const entity = this.bundle.messages.get(id);
+
+    if (!entity) {
+      return [id, [new L10nError(`Unknown entity: ${id}`)]];
     }
 
-    return this.fetch(langs.slice(1)).then(
-      nextLangs => this._resolve(nextLangs, keys, formatter, resolved));
+    return this.bundle.format(entity, args);
   }
 
-  formatEntities(...keys) {
-    return this.fetch().then(
-      langs => this._resolve(langs, keys, this._formatEntity));
-  }
+  formatEntity(id, args) {
+    const entity = this.bundle.messages.get(id);
 
-  formatValues(...keys) {
-    return this.fetch().then(
-      langs => this._resolve(langs, keys, this._formatValue));
-  }
+    if (!entity)  {
+      return [
+        { value: id, attrs: null },
+        [new L10nError(`Unknown entity: ${id}`)]
+      ];
+    }
 
-  _getEntity(lang, name) {
-    const cache = this.env.resCache;
+    const [value] = this.bundle.format(entity, args);
 
-    // Look for `name` in every resource in order.
-    for (let i = 0, resId; resId = this.resIds[i]; i++) {
-      const resource = cache.get(resId + lang.code + lang.src);
-      if (resource instanceof L10nError) {
-        continue;
-      }
-      if (name in resource) {
-        return resource[name];
+    const formatted = {
+      value,
+      attrs: null,
+    };
+
+    if (entity.traits) {
+      formatted.attrs = Object.create(null);
+      for (let trait of entity.traits) {
+        const [attrValue] = this.bundle.format(trait, args);
+        formatted.attrs[trait.key.name] = attrValue;
       }
     }
-    return undefined;
-  }
 
-  _memoizeIntlObject(ctor, {code}, opts) {
-    const cache = IntlObjects.get(ctor) || {};
-    const id = code + JSON.stringify(opts);
+    // XXX return errors
+    return [formatted, []];
+  }
 
-    if (!cache[id]) {
-      cache[id] = new ctor(code, opts);
-      IntlObjects.set(ctor, cache);
-    }
 
-    return cache[id];
+  formatValues(...keys) {
+    return this._formatKeys(keys, SimpleContext.prototype.formatValue);
```

Comment:
I think it has to be `this.constructor.prototype.formatValue` but yeah, good call.